docs: expand documentation about testing (#5346)
restic/restic: doc/090_participating.rst

In case you want or need to create tests for an enhancement or a new feature of restic,
here is a brief description of how to write tests.

Tests are typically falling into two categories: functional tests (unit tests) and integration tests.
Functional tests will verify the correct workings of a function or a set of functions.
See more on integration tests below.

The restic test package located in ``internal/test``, here named ``rtest``,
provides the following very basic test functions:
::

 rtest.Equals(t, a, b, msg)         compares two values, fails test if a != b, optional ``msg`` string
                                    for detailed message
 rtest.Assert(t, a == b, "msg", more variables a, b) checks for a condition to be true,
                                    <msg> is a format string to represent the values of <a and b>
 rtest.OK(t, err)                   expects err to be ``nil``, otherwise fails test
 rtest.OKs(t, errs)                 expects a slice of errs to be ``nil``

Functional tests
================

The packages in ``internal/...`` often provide ``Test*(...)`` functions and structs or
have a dedicated test package like ``internal/backend/test``.

> A good starting points are also the testing.go files that exist in several places.

Reviewer: ```suggestion
A good starting points are also the `testing.go` files that exist in several places.
```

Functional tests are stored in the same directory as their function, e.g.
``internal/<sub-component>/<function>_test.go``.

Tests in ``internal`` that need a full repository should just create one using the
memory backend by calling ``repository.TestRepository(t)``.

``checker.TestCheckRepo()`` can be used to verify the repository integrity.

In general, in memory operation should be preferred over creating temporary files.

> If necessary, store them in ``t.TempDir()``. However, in most cases only a

Reviewer: ```suggestion
If necessary, temporary files can be stored in ``t.TempDir()``. However, in most cases only a
```

> few methods provided by the Repository are required.

Author: done.

> Then some helper like ``TestTreeMap`` can be used or just create a basic mock yourself.

Reviewer: ```suggestion
Then some helper like ``data.TestTreeMap`` can be used or just create a basic mock yourself.
```

For backends the test suite in ``internal/backend/test`` is mandatory.

For feature flags use ``defer feature.TestSetFlag(t, feature.Flag, feature.DeviceIDForHardlinks, true)()``

> to enable them temporarily in tests.

Reviewer: ```suggestion
To temporarily enable feature flags in tests, use ``defer feature.TestSetFlag(t, feature.Flag, feature.DeviceIDForHardlinks, true)()``.
```

Such tests must not run in parallel as this changes global state.

Integration tests
=================

The classical helpers for integration tests are, amongst others:

- ``env, cleanup := withTestEnvironment(t)``: build an environment for tests
- ``datafile := testSetupBackupData(t, env)``: initialize a repo, unpack standard backup tree structure
- ``testRunBackup(t, "", []string{env.testdata}, BackupOptions{}, env.gopts)``: backup all of the standard tree structure
- ``testListSnapshots(t, env.gopts, <n>)``: check that there are <n> snapshots in the repository
- ``testRunCheck(t, env.gopts)``: check that the repository is sound and happy

> - the above mentioned ``rtest.OK()``, ``rtest.Equals()``, ``rtest.Assert()````

Reviewer: ```suggestion
- the above mentioned ``rtest.OK()``, ``rtest.Equals()``, ``rtest.Assert()`` helpers
```

Integration tests however test overall workings of a command:
::

 run a ``backup``, compare number of files backup with the expected number of files
 run a ``backup``, run the ``ls`` command with a ``sort`` option
 compare actual output with the expected output.

For all backup related functions there is a directory tree which will be backed up
by default, to be found at ``cmd/restic/testdata/backup-data.tar.gz``.
In this compressed directory tree in which you will find files, hardlinked files,
symlinked files, an empty directory and a simple directory structure which is ok for testing purposes.

Have a look at the ``cmd/restic/*_integration_test.go`` files. In there you will find a
wealth of helper functions which make life easier for you when you are testing.

These are the standard functions to check for a specific type of result
::

 rtest.Equals(t, a, b, "msg")       compares two values, fail test if differ
 rtest.Assert(t, a == b, "msg")     checks for a condition to be true
 rtest.OK(t, err)                   expects err to be ``nil``, otherwise fail test
 rtest.OKs(t, errs)                 expects a slice of errs to be ``nil``

>  testListSnapshots(t, env.gopts, n) expects exactly <n> snapshots in the repository

Reviewer: rtest and testListSnapshots should be explained separately.

If you need to capture output from a command, look out for examples with ``withCaptureStdout`` in
``cmd/restic/*_test.go``. Basicly, stdout gets attached to a ``bytes.Buffer``.

> In addition, you can use the global option ``gopts.JSON: true`` to generate JSON output

Reviewer: ```suggestion
In addition, you can set the option ``gopts.JSON: true`` to generate JSON output
```

Even though gopts is called global options, it's not global in test cases.

This is a typical setup for a backup / find scenario is
::

 env, cleanup := withTestEnvironment(t)
 defer cleanup()

 testSetupBackupData(t, env)
 opts := BackupOptions{}
 testRunBackup(t, env.testdata+"/0", []string{"."}, opts, env.gopts)
 testListSnapshots(t, env.gopts, 1)

 results = testRunFind(t, false, FindOptions{}, env.gopts, "testfile")
 lines := strings.Split(string(results), "\n")
 rtest.Assert(t, len(lines) == 2, "expected one file, found (%v) in repo", len(lines)-1)

commented
::

 env, cleanup := withTestEnvironment(t)        // setup test
 testSetupBackupData(t, env)                   // init repository
 testRunBackup(t, env.testdata+"/0", ...)      // run a backup
 testListSnapshots(t, env.gopts, 1)            // make sure you have one snapshot
 testRunFind(t, false, FindOptions{}, ...)     // run restic find, look for file 'testfile'
 lines := strings.Split(string(results), "\n") // there is always a ``\n`` at  the end of the output!
 rtest.Assert(t, len(lines) == 2, ...)         // make sure that you have correct output

When it comes to writing tests in the ``internal`` directory tree, things can become

> a bit more tricky, since you haven't necessarily got the full support of an opened repository

Reviewer: That's the whole point and in many cases is perfectly fine. See https://github.com/restic/restic/pull/5346/changes#r2751469634 regarding what to use instead.
